Sources/Util/Heap.swift

```diff
@@ -19,232 +19,176 @@
 
 import Foundation
 
-struct Heap<T: Comparable> {
-    /** The array that stores the heap's nodes. */
-    var nodes = [T]()
-
-    /**
-     * Determines how to compare two nodes in the heap.
-     * Use '>' for a max-heap or '<' for a min-heap,
-     * or provide a comparing method if the heap is made
-     * of custom elements, for example tuples.
-     */
-    private var orderCriteria: (T, T) -> Bool
+/**
+ * `HeapNode` is a node of `Heap`.
+ */
+struct HeapNode<K, V> {
+    fileprivate let key: K
+    let value: V
 
-    /// Create a max-heap
-    init() {
-        self.orderCriteria = { left, right -> Bool in
-            left > right
-        }
+    init(key: K, value: V) {
+        self.key = key
+        self.value = value
     }
+}
 
-    /// Create a max-heap
-    init(array: [T]) {
-        self.orderCriteria = { left, right -> Bool in
-            left > right
-        }
-        self.configureHeap(from: array)
-    }
+/**
+ * `Heap` is a heap implemented with max heap.
+ */
+class Heap<K: Comparable, V: Equatable>: Sequence, IteratorProtocol {
+    private var nodes: [HeapNode<K, V>] = []
 
     /**
-     * Creates an empty heap.
-     * The sort function determines whether this is a min-heap or max-heap.
-     * For comparable data types, > makes a max-heap, < makes a min-heap.
+     * `peek` returns the maximum element from this Heap.
      */
-    init(sort: @escaping (T, T) -> Bool) {
-        self.orderCriteria = sort
-    }
+    func peek() -> HeapNode<K, V>? {
+        guard self.nodes.isEmpty == false else {
+            return nil
+        }
 
-    /**
-     * Creates a heap from an array. The order of the array does not matter;
-     * the elements are inserted into the heap in the order determined by the
-     * sort function. For comparable data types, '>' makes a max-heap,
-     * '<' makes a min-heap.
-     */
-    init(array: [T], sort: @escaping (T, T) -> Bool) {
-        self.orderCriteria = sort
-        self.configureHeap(from: array)
+        return self.nodes[0]
     }
 
     /**
-     * Configures the max-heap or min-heap from an array, in a bottom-up manner.
-     * Performance: This runs pretty much in O(n).
+     * `length` is the number of elements in this Heap.
      */
-    private mutating func configureHeap(from array: [T]) {
-        self.nodes = array
-        for index in stride(from: self.nodes.count / 2 - 1, through: 0, by: -1) {
-            self.shiftDown(index)
-        }
+    func length() -> Int {
+        return self.nodes.count
     }
 
     var isEmpty: Bool {
-        return self.nodes.isEmpty
-    }
-
-    var count: Int {
-        return self.nodes.count
+        return self.nodes.count == 0
```

Comment:
```suggestion
        return self.nodes.isEmpty
```

count may slow than isEmpty

Reply:
That's right. I modified it.